sa/saro.go

```diff
@@ -861,10 +863,17 @@ func (ssa *SQLStorageAuthorityRO) CountPendingAuthorizations2(ctx context.Contex
 	return &sapb.Count{Count: count}, nil
 }
 
-// GetValidOrderAuthorizations2 is used to find the valid, unexpired authorizations
-// associated with a specific order and account ID.
+// GetValidOrderAuthorizations2 is used to get all authorizations
+// associated with the given Order ID.
+// NOTE: The name is outdated. It does *not* filter out invalid or expired
+// authorizations; that it left to the caller. It also ignores the RegID field
+// of the input: ensuring that the returned authorizations match the same RegID
+// as the Order is also left to the caller. This is because the caller is
+// generally in a better position to provide insightful error messages, whereas
+// simply omitting an authz from this method's response would leave the caller
+// wondering why that authz was omitted.
```

Comment:
Why is the caller in a better position to provide an insightful error message? Why can't we do most of the same filtering here that we now do in the RA?

I'm not terribly happy with a method whose name implies a contract that it's doc comment and implementation contradict. Do we have a plan for resolving this in the future?

Reply:
Agreed, I don't love it either. I'm happy to do follow-up changes that create an identical SA method with a better name, switch the RA over to call that method instead, and then remove this one. But I didn't want to mix that kind of very mechanical change with the logic changes present in the rest of this PR, I thought it would make it too sprawling and confusing. Happy to do it anyway if you prefer it that way.

I think it is more appropriate for callers of this function to do the filtering, because they have to do filtering _anyway_, e.g. based on things like "is the challenge type that was used to fulfill this authz still enabled?".

Also, imagine this hypothetical: the RA is working with an order for 5 names, it asks the SA for all "valid" authzs associated with that order, and it gets 4 results back. What is it supposed to do with that information? It has no idea why it got too few results -- was the fifth one expired? invalid? associated with the wrong regID? It's much better for the RA to get all of the results, and then handle them as a full set.

Reply:
> Also, imagine this hypothetical: the RA is working with an order for 5 names, it asks the SA for all "valid" authzs associated with that order, and it gets 4 results back. What is it supposed to do with that information? It has no idea why it got too few results -- was the fifth one expired? invalid? associated with the wrong regID? It's much better for the RA to get all of the results, and then handle them as a full set.

Would it really get back 4 authzs? It seems like the best call would be to return an error concerning the fifth authz.

Reply:
Yes, the previous implementation of this function would have returned 4 authzs. That's exactly why I've removed the SA-level filtering in this PR.

Reply:
Oh, that's totally understandable. As an alternative, I'm suggesting that we filter out invalid authorizations and return an error if the remaining valid authorizations are insufficient to fulfill the request rather than leaving that to the RA. There's still a small amount of filtering to be done in the RA, but I think that could be more efficient than sending RPCs full of invalid authorizations back to the RA.

Reply:
Summary of what we discussed offline for posterity:

The motivation for this PR is to address performance issues with the underperforming index on the status column of the authz2 table. By simplifying the query and shifting the filtering directly into Go, we can completely bypass this issue. We anticipate that any performance degradation resulting from this change will be negligible. This method is invoked only once (during finalization), and the number of results is constrained by the limit on names allowed in an order, which currently stands at 100. Typically, most requests will return one or two authorizations, and those containing a larger number are usually valid, based on our observations.

We considered maintaining this method's contract and even improving filtering in the SA by adjusting the request to include enabled challenge types. However, concerns were raised about the uncertainty of future filtering requirements. Continuously updating the request fields for this method could become problematic. Therefore, we've decided to limit filtering strictly to the RA and accept the additional RPC time.